source/extensions/filters/common/rbac/utility.h

```diff
@@ -32,6 +32,30 @@ namespace RBAC {
 struct RoleBasedAccessControlFilterStats {
   ENFORCE_RBAC_FILTER_STATS(GENERATE_COUNTER_STRUCT)
   SHADOW_RBAC_FILTER_STATS(GENERATE_COUNTER_STRUCT)
+
+  Stats::Scope& scope_;
+  Stats::StatName per_policy_stat_;
+  Stats::StatName per_policy_shadow_stat_;
+  Stats::StatNameDynamicPool pool_;
+
+  void inc_policy_allowed(absl::string_view name) {
+    incCounter(per_policy_stat_, name, ".allowed");
+  }
+
+  void inc_policy_denied(absl::string_view name) { incCounter(per_policy_stat_, name, ".denied"); }
+
+  void inc_policy_shadow_allowed(absl::string_view name) {
+    incCounter(per_policy_shadow_stat_, name, ".shadow_allowed");
+  }
+
+  void inc_policy_shadow_denied(absl::string_view name) {
+    incCounter(per_policy_shadow_stat_, name, ".shadow_denied");
+  }
+
+  void incCounter(const Stats::StatName& prefix, absl::string_view name, absl::string_view suffix) {
+    Stats::StatName metric_name = pool_.add(absl::StrCat(name, suffix));
```

Comment:
I think this leaks memory. The pool will grow indefinitely every time counter is incremented.

Reply:
I was following the dynamodb filter as an example of registering stats at request time https://github.com/envoyproxy/envoy/blob/530e8a1a6e32f9c9b3fd09008bb4cdb540f59eb5/contrib/dynamo/filters/http/source/dynamo_filter.cc#L175 - however, now I am looking into a separate bug that seems to be present when using `StatNameDynamicPool` in our code (failed assertion in `symbol_table.cc`). Since the stat names can be derived from the filter config / route-specific filter config I think `StatNameSet` might be more appropriate here based on https://github.com/envoyproxy/envoy/blob/main/source/docs/stats.md

Reply:
Yes, stat names need to be allocated at the configuration time ideally. That would work better here.

Reply:
I am planning to put a `Filters::Common::RBAC::RoleBasedAccessControlFilterStats` in the `RoleBasedAccessControlRouteSpecificFilterConfig` so we can register stats at configuration time but I'm not sure how to fetch the http connection manager's stat prefix from that class. To create a stats obj we need:

* a stats prefix
* the protobuf config
* stats scope

The `RoleBasedAccessControlFilterConfig` class has access to those components from the http filter factory, but the `RoleBasedAccessControlRouteSpecificFilterConfig` only has access to the protobuf config and a stats scope, https://github.com/DataDog/envoy/blob/ce4c15a284296fac818b737537e54f253c43fc0e/source/extensions/filters/http/rbac/config.cc#L28. Would you have any pointers on how to get the filter's stat prefix in the route-specific filter config @kyessenov @yanavlasov?

Reply:
You may need to do 2-phase initialization for stats. You can use the `ServerFactoryContext::scope()` function but the prefix I think would need to be stored in the factory and passed to the route specific config to initialize stats.

Reply:
I updated the PR to use stats at config initialization time - @yanavlasov would you be ok to defer support for per-rule stats from the per-route filter to another PR? Right now if using the rbac-per-route + per-policy stats, the metric would be "unknown". It's not immediately clear to me how we'd store the prefix coming from the listener filter config in `ServerFactoryContext` and how to retrieve it from the route specific config's factory function (i.e. mapping from route back to the listener's stat prefix)

Reply:
@yanavlasov @kyessenov I changed it back to use the `StatNameDynamicPool` and updated how we use it (creating the pool per request like the dynamodb filter does, similar to https://github.com/envoyproxy/envoy/blob/63bc9b564b1a76a22a0d029bcac35abeffff2a61/contrib/dynamo/filters/http/source/dynamo_filter.cc#L175). I confirmed there's no memory leak there after running loadtests in our staging env and checking memory usage.

> You can use the ServerFactoryContext::scope() function but the prefix I think would need to be stored in the factory and passed to the route specific config to initialize stats.

I believe the factory is shared among listeners based on the comment here https://github.com/envoyproxy/envoy/blob/63bc9b564b1a76a22a0d029bcac35abeffff2a61/envoy/server/factory_context.h#L144 so I don't see how we could store the stat prefix there.